File: irrgarten java/src/irrgarten/LabyrinthCharacter.java

package irrgarten;

/**
 *
 * @author laurazafra
 */
public abstract class LabyrinthCharacter {
    //si la clase es abstracta tiene que tener algún método abstracto
    private String name;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;
    
    public LabyrinthCharacter (String n, float i, float s, float h){
        name = n;
        intelligence = i;
        strength = s;
        health = h;
        row = -1;
        col = -1;
    }
    
    public LabyrinthCharacter (LabyrinthCharacter other){
        name = other.name;
        intelligence = other.intelligence;
        strength = other.strength;
        row = other.row;
        col = other.col;
        health = other.health;
    }
    
    //Devuelve true si el monstruo está muerto
    public boolean dead (){
        return (health <= 0);
    }
    
    //METODOS PARA DEVOLVER VALOR DE LA VARIALE
    public int getRow(){
        return row;
    }
    public int getCol(){
        return col;
    }

    //Nuevas de la practica 4
    protected float getIntelligence(){
        return intelligence;
    }

    protected float getStrength(){
        return strength;
    }

    protected float getHealth(){
        return health;
    }

    protected void setHealth(float s){
        health=s;
    }
    
    //Es un modificador en una única llamada de los atributos row y col.
    public void setPos (int row, int col){
        this.row = row;
        this.col = col;
    }
    
    //Genera un representación del estado completo del monstruo en forma de cadena de caracteres.
    public String toString (){
        return "Name: " + name + "\n" +
                "Intelligence: " + intelligence + "\n" +
                "Strength: " + strength + "\n" +
                "Health: " + health + "\n" +
                "Row: " + row + "\n" +
                "Col: " + col + "\n";
    }
    
    //decrementa en una unidad el atributo que representa la salud del monstruo.
    protected void gotWounded(){
        if (health > 0) health--;
    }
    
    public abstract float attack();
    
    public abstract boolean defend(float attack);
}

File: Irrganten rb/irrgarten/LabyrinthCharacter.rb
# frozen_string_literal: true
require_relative 'Dice'
require_relative 'Weapon'
require_relative 'Shield'

module Irrgarten
  class LabyrinthCharacter
    attr_accessor :row
    attr_accessor :col
    attr_accessor :health         #permitiendo leer su valor desde dentro de la clase o sus subclases.
    attr_reader   :intelligence      #Sin embargo, intentar acceder al atributo desde fuera de la clase generará un error
    attr_reader   :strength
    attr_accessor :name

    #Metodos
    #Constructor
    def initialize(n, i, s, h)
      @name=n
      @intelligence=i
      @strength=s
      @health = h
      @row=-1
      @col=-1
    end

    def copy(other)
      @name = other.name
      @intelligence = other.intelligence
      @strength = other.strength
      @health = other.health
      @row= other.row
      @col= other.col
    end

    #Devuelve true si está muerto.
    def dead
      return (@health<=0)
    end

    #El GETROW() Y GETCOL(), estos consultares estan impicitos con lo que se ha puesto al principio de ' attr_accessor '
    # Igual para GETHELATH(), GETSTRENGT() y GETINTELLIGENCE() y SETHEALTH()

    #Es un modificador en una única llamada de los atributos row y col.
    def set_pos (row, col)
      @row = row
      @col = col
    end

    def to_string
      return "Name: " + @name + "\n" +
        "Intelligence: " + @intelligence.to_s + "\n" +
        "Strength: " + @strength.to_s + "\n" +
        "Health: " + @health.to_s + "\n" +
        "Row: " + @row.to_s + "\n" +
        "Col: " + @col.to_s + "\n"
    end

    #Este método decrementa en una unidad el atributo que representa la salud
    protected def got_wounded
      if (@health > 0) then
        @health=@health-1
      end
    end

    #métodos abstractos no se declaran: attack() y defend()

    #para los metodos abstractps
    private_class_method :new

  end
end
